rules/manager.go

```diff
@@ -407,12 +422,133 @@ func (g *Group) Eval(ctx context.Context, ts time.Time) {
 	}
 }
 
+// RestoreForState restores the 'for' state of the alerts
+// by looking up last ActiveAt from storage.
+func (g *Group) RestoreForState(ts time.Time) {
+	maxtMS := int64(model.TimeFromUnixNano(ts.UnixNano()))
+	// We allow restoration only if alerts were active before some certain time.
+	mint := ts.Add(-g.opts.OutageTolerance)
+	mintMS := int64(model.TimeFromUnixNano(mint.UnixNano()))
+	q, err := g.opts.TSDB.Querier(g.opts.Context, mintMS, maxtMS)
+	if err != nil {
+		level.Error(g.logger).Log("msg", "Failed to get Querier", "err", err)
+	}
+
+	for _, rule := range g.Rules() {
+		alertRule, ok := rule.(*AlertingRule)
+		if !ok {
+			continue
+		}
+
+		alertHoldDuration := alertRule.HoldDuration()
+
+		alertRule.ForEachActiveAlert(func(a *Alert) {
+
+			smpl := alertRule.forStateSample(a, time.Now(), 0)
+			var matchers []*labels.Matcher
+			for _, l := range smpl.Metric {
+				mt, _ := labels.NewMatcher(labels.MatchEqual, l.Name, l.Value)
+				matchers = append(matchers, mt)
+			}
+
+			sset, err := q.Select(nil, matchers...)
+			if err != nil {
+				level.Error(g.logger).Log("msg", "Failed to restore 'for' state",
+					labels.AlertName, alertRule.Name(), "stage", "Select", "err", err)
+				return
+			}
+
+			seriesFound := false
+			var s storage.Series
+			for sset.Next() {
+				if len(sset.At().Labels()) == len(smpl.Metric) {
+					s = sset.At()
+					seriesFound = true
+					break
+				}
+			}
+
+			if !seriesFound {
+				return
+			}
+
+			// Series found for the 'for' state.
+			var v float64
+			var t int64
+			it := s.Iterator()
+			for it.Next() {
+				t, v = it.At()
+			}
+			if value.IsStaleNaN(v) { // Alert was not active.
+				return
+			}
+
+			// v = the latest ActiveAt of the alert.
+			restoredActiveAt := time.Unix(int64(v), 0)
+			crashTime := time.Unix(t/1000, 0)
+			if crashTime.Sub(restoredActiveAt) >= alertHoldDuration {
+				// The alert was firing when it crashed. Hence we need not wait for grace
+				// period, we can immediatly make it fire.
```

Comment:
I think we should still apply the same logic here, and wait the 10m. Just because it was firing an hour ago doesn't mean the condition is still happening.

Reply:
Okay. In that case we have no use of timestamp of the sample, as we wait for 10 minutes in any case.